pkg/controller/job/job_controller.go

```diff
@@ -1554,13 +1569,18 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 	return calculated
 }
 
-// countValidPodsWithFilter returns number of valid pods that pass the filter.
-// Pods are valid if they have a finalizer and, for Indexed Jobs, a valid
-// completion index.
-func countValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) int {
-	result := len(uncounted)
+// getValidPodsWithFilter returns the valid pods that pass the filter.
+// Pods are valid if they have a finalizer or in uncounted set
+// and, for Indexed Jobs, a valid completion index.
+func getValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) []*v1.Pod {
+	var result []*v1.Pod
 	for _, p := range pods {
 		uid := string(p.UID)
+
+		if uncounted.Has(uid) {
+			result = append(result, p)
+		}
+
 		// Pods that don't have a completion finalizer are in the uncounted set or
 		// have already been accounted for in the Job status.
 		if uncounted != nil && (!hasJobTrackingFinalizer(p) || uncounted.Has(uid) || expectedRmFinalizers.Has(uid)) {
```

Comment:
@kannon92 this `uncounted != nil` should have gone away, as it's always true, right?

Reply:
Ah sorry.  I'll open up a PR for that.

Reply:
Actually I'm not sure..

```
func getStatus(job *batch.Job, pods []*v1.Pod, uncounted *uncountedTerminatedPods, expectedRmFinalizers sets.String) (succeeded, failed int32) {
	succeeded = job.Status.Succeeded
	failed = job.Status.Failed
	succeeded += int32(countValidPodsWithFilter(job, pods, uncounted.Succeeded(), expectedRmFinalizers, func(p *v1.Pod) bool {
		return p.Status.Phase == v1.PodSucceeded
	}))
	failed += int32(countValidPodsWithFilter(job, pods, uncounted.Failed(), expectedRmFinalizers, func(p *v1.Pod) bool {
		if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && job.Spec.PodFailurePolicy != nil {
			if !isPodFailed(p, job) {
				return false
			}
			_, countFailed, _ := matchPodFailurePolicy(job.Spec.PodFailurePolicy, p)
			return countFailed
		} else {
			return isPodFailed(p, job)
		}
	}))
	return succeeded, failed
}
```

Is it not possible for `uncounted.Failed()` or `uncounted.Succeeded()` to be nil?

Reply:
I think not for jobs tracked with finalizers, which is now all.

Reply:
https://github.com/kubernetes/kubernetes/pull/115970

Reply:
Any action required from me here?

Reply:
Nope!

Reply:
Probably a rebase if you haven’t done that yet.